### PR TITLE
Refactoring the core algorithmic part with several bug fixes

### PR DESCRIPTION
**Pull request by kibaekkim**

### CODE REVIEW

**Comment by kibaekkim** [comment]
The next step is to resolve #35

**Comment by kibaekkim** [comment]
```
Test Summary:                           | Pass  Fail  Total
MathOptInterface                        |   85    23    108
  SolverName                            |    1            1
  supports_default_copy_to              |    2            2
  Unit                                  |   82    23    105
    solve_qcp_edge_cases                |   12     2     14
      Duplicate on-diagonal             |    7            7
      Duplicate off-diagonal            |    5     2      7
    add_variable                        |    3            3
    solve_duplicate_terms_vector_affine |          8      8
    solve_single_variable_dual_max      |    3     1      4
    solve_single_variable_dual_min      |    3     1      4
    solver_name                         |    1            1
    solve_time                          |    4     2      6
    solve_duplicate_terms_obj           |    9     1     10
    raw_status_string                   |    6            6
    add_variables                       |    3            3
    feasibility_sense                   |    3            3
    max_sense                           |    3            3
    silent                              |    4            4
    min_sense                           |    3            3
    solve_duplicate_terms_scalar_affine |          8      8
    time_limit_sec                      |    4            4
    solve_qp_edge_cases                 |   21           21
```

**Comment by kibaekkim** [comment]
```
Test Summary:                           | Pass  Fail  Total
MathOptInterface                        |  105     3    108
  SolverName                            |    1            1
  supports_default_copy_to              |    2            2
  Unit                                  |  102     3    105
    solve_qcp_edge_cases                |   14           14
    add_variable                        |    3            3
    solve_duplicate_terms_vector_affine |    8            8
    solve_single_variable_dual_max      |    3     1      4
    solve_single_variable_dual_min      |    3     1      4
    solver_name                         |    1            1
    solve_time                          |    6            6
    solve_duplicate_terms_obj           |    9     1     10
    raw_status_string                   |    6            6
    add_variables                       |    3            3
    feasibility_sense                   |    3            3
    max_sense                           |    3            3
    silent                              |    4            4
    min_sense                           |    3            3
    solve_duplicate_terms_scalar_affine |    8            8
    time_limit_sec                      |    4            4
    solve_qp_edge_cases                 |   21           21
```

File: src/lp_opt.jl [hunk 1]
```diff
@@ -178,5 +154,27 @@ function solve_lp(c_init,A,b,x_L,x_U,constraint_lb,constraint_ub,mu,x_hat,Δ)
...
+	return Xsol, lambda, status
+end
+
+function solve_lp(env::SLP, Δ)
+
+	A = spzeros(env.problem.m, env.problem.n)
+	for Ai = 1:length(env.problem.j_str)
+		# TODO: This might cause problem in terms of lambda
+		# KK: WHY?
+		A[env.problem.j_str[Ai][1], env.problem.j_str[Ai][2]] = env.dE[Ai]
```
[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/lp_opt.jl]: Although the lines commented are outdated, the dimension of Jacobian matrix `A` should be fixed; that is, determined by the problem. Even if some rows or columns become zeros, that should not be affected.

There was an issue related to the dimension of lambda. This must be independent to this. So far, after fixing several bugs, I do not see the issue anymore though.

File: src/SLP_line_search.jl [hunk 2]
```diff
@@ -288,3 +287,215 @@ function SLP_line_search(model::NloptProblem)
...
+        elseif env.problem.x_U[i] < Inf
+            env.x[i] = env.problem.x_U[i]
+        else
+            env.x[i] = 0.0
+        end
+    end
+
+    for i = 1:env.problem.m
+        if env.problem.g_U[i] < Inf && env.problem.g_L[i] > -Inf
+            @error "Range constraints are not supposed."
```
[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/SLP_line_search.jl:356]: I don't want the algorithm to just run without knowing whether range constraints are properly addressed or now. So far I cannot find any clue that is the case.

File: src/lp_opt.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ssadat) [reply src/lp_opt.jl]: Ok I will create a separate branch and work on fixing this issue

File: src/SLP_line_search.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssadat) [reply src/SLP_line_search.jl:356]: I think line `106` of `lp_opt.jl` handles it, doesn’t it?
I might need to removed  `&& constraint_lb[i] < constraint_ub[i]` from the `elseif`

[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/SLP_line_search.jl:356]: Other things may also need to be done: multipliers, violations, and so on.

File: src/lp_opt.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (kibaekkim) [reply src/lp_opt.jl]: Which issue? This part is not an issue anymore.

[the remark this answers is not in the document]
Comment (ssadat) [reply src/lp_opt.jl]: To make sure that this sparsity doesn't cause lambda dimension discrepancy

[the remark this answers is not in the document]
Comment (ssadat) [reply src/lp_opt.jl]: Or if you think for sure this won't cause a problem then I won't change it

**Comment by kibaekkim** [comment]
```
Test Summary:    | Pass  Total
MathOptInterface |  108    108
    Testing ActiveSetMethods tests passed
```

**Pull request merged**